### ISSUE TITLE
Use undefined as Tensor of zeroes in the python Function API

### ISSUE THREAD

**Issue by albanD** [pytorch/pytorch#41359]
We now fully support undefined Tensors in the backward pass to represent a Tensor full of zeroes.
We should leverage this and stop creating full size Tensors for custom Function (both on python and cpp side)

cc @ezyang @SsnL @albanD @zou3519 @gqchen @yf225 @glaringlee @VitalyFedyunin @ngimel

**Comment by albanD** [comment]
More details of the two part that are covered by this issue.

1. Stop materializing outputs
For this part, we just want to make sure that `None` or `undefined` Tensors returned by the custom backward are passed as `undefined` to the rest of the computations.
The only BC breaking change this leads to is when this output is directly linked to either an `AccumulateGrad` (and so the .grad field of a leaf Tensor) or an input Tensor from `autograd.grad()`. In this case, the user will get None/undefined for Tensors that used to get a Tensor full of 0s.
This is properly handled by our optimizers and will cause clear errors for users that expect a Tensor here.
We already use None, an error or a Tensor full of 0s to represent a Tensor full of 0s and which one the user get depends on the internal implementation.

2. Stop materializing the inputs (can be done in a second PR)
This one would mean that we stop creating Tensor full of 0s and give undefined Tensor in cpp and None in python.
This one is more problematic from a BC point of view because it means that all the user's custom Function need to handle such special inputs.
Given that this is only for performance in very specific case, we can add this as an opt-in feature. So the behavior wouldn't change for current users but users that define a static attribute `dont_materialize_grads = True` will get the new behavior and get None/undefined gradients that they need to properly handle.

**Comment by SsnL** [comment]
> This is properly handled by our optimizers and will cause clear errors for users that expect a Tensor here.
We already use None, an error or a Tensor full of 0s to represent a Tensor full of 0s and which one the user get depends on the internal implementation.

I just want to say that this is **not** properly handled by the optimizers. For the ones doing weight decay and/or momentum, having `None` gradient vs having `zero` gradient do different things. It is use to decide when a variable starts to accept momentum / weight decay. I would agree with the argument that this is a somewhat weird behavior though.

**Comment by albanD** [comment]
> I just want to say that this is not properly handled by the optimizers.

Good point. It is handled but not the exact same way.

My reasoning is:
- We don't guarantee this behavior (you can get None, an error or a Tensor full of 0s already). So the behavior for inputs independent of the output is already in a bad state.
- This will affect a very small number of users
- A user returning None from their backward will get None for their .grad field which feels consistent
- The performance improvement can be significant depending on the tasks (following Node will also skip computations)
- Drawback: Some user custom Functions that are independent of the input (from the grad point of view, not forward point of view) will now return None gradient instead of a Tensor full of 0s.
  - This can confuse the user for that used to get a Tensor full of 0s. But I would argue that None is more consistent if their custom function returned None.
  - This will change the optimizer behavior.

Do you agree that this BC-breaking change is worth it?

**Comment by SsnL** [comment]
@albanD It would break some reproducibility silently though, which is my main concern. I think we should more properly define what to do with momentum/decay when grad is `None` and zeros. Also another thing to think about is whether we want to change `.zero_grad()` to set `None` instead.

**Comment by albanD** [comment]
> @albanD It would break some reproducibility silently though, which is my main concern. 

Yes we marked it as BC-breaking.
Given that it only affects people that will return `None` for a Tensor that requires grad in a custom function, I think this is ok. Mainly because very few people actually do that, or are even aware that they are allowed to return `None` for inputs that require gradients.
If you feel very strongly about this, we can make this materialization of the output conditional on the same flag as https://github.com/pytorch/pytorch/pull/41490 ? And that way it is only an opt-in feature that won't change the default behavior. You think that would be better?

>  Also another thing to think about is whether we want to change .zero_grad() to set None instead.

This is discussed there: https://github.com/pytorch/pytorch/issues/41696
I'm not sure we really want to change the default. But giving the user both options very clearly sounds good.
Please do send any feedback you have there!
